src/System/IO/ByteBuffer.hs

```diff
@@ -92,10 +104,53 @@ data BBRef = BBRef {
       -- the 'ByteBuffer'
     }
 
-type ByteBuffer = IORef BBRef
+-- | Exception that is thrown when an invalid 'ByteBuffer' is being used that is no longer valid.
+--
+-- A 'ByteBuffer' is considered to be invalid if
+--
+-- - it has explicitly been freed
+-- - an Exception has occured during an operation that modified it
+data ByteBufferException = ByteBufferException
+    { _bbeLocation :: !String
+      -- ^ function name that caused the exception
+    , _bbeException :: !String
+      -- ^ printed representation of the exception
+    }
+    deriving (Typeable, Eq)
+instance Show ByteBufferException where
+    show (ByteBufferException loc e) = concat
+        ["ByteBufferException: ByteBuffer was invalidated because of Exception thrown in "
+        , loc , ": ", e]
+instance Exception ByteBufferException
+
+type ByteBuffer = IORef (Either ByteBufferException BBRef)
+
+-- | On any Exception, this will invalidate the ByteBuffer and re-throw the Exception.
+--
+-- Invalidating the 'ByteBuffer' includes freeing the underlying pointer.
+bbHandler ::
```

Comment:
OK, one last complaint: `bbHandler` is also only used in conjunction with `withBBRef`, so I'd combine that too to get the nicer type signature

```
bbHandler ::
     String
  -> ByteBuffer
  -> (BBRef -> IO a)
  -> IO a
```